syscall_sdl.c

```diff
@@ -9,9 +9,73 @@
 
 #include "state.h"
 
+#define MAX_EVENT_QUEUE_SIZE 100
+
+enum {
+    KEY_EVENT = 0,
+    MOUSE_MOTION_EVENT = 1,
+    MOUSE_BUTTON_EVENT = 2,
+};
+
+typedef struct {
+    uint32_t keycode;
+    uint8_t state;
+} key_event_t;
+
+typedef struct {
+    int32_t xrel;
+    int32_t yrel;
+} mouse_motion_t;
+
+typedef struct {
+    uint8_t button;
+    uint8_t state;
+} mouse_button_t;
+
+typedef struct {
+    uint32_t type;
+    union {
+        key_event_t key_event;
+        mouse_motion_t mouse_motion;
+        mouse_button_t mouse_button;
+    };
+} event_t;
+
+typedef struct {
+    event_t events[MAX_EVENT_QUEUE_SIZE];
+    size_t start;
+    size_t end;
+    bool full;
+} event_queue_t;
+
 static SDL_Window *window = NULL;
 static SDL_Renderer *renderer;
 static SDL_Texture *texture;
+static event_queue_t event_queue = {
+    .events = {},
+    .start = 0,
+    .end = 0,
+    .full = false,
+};
+
+static bool event_pop(event_t *event)
+{
+    if (event_queue.start == event_queue.end)
+        return false;
+    *event = event_queue.events[event_queue.start++];
+    event_queue.start %= MAX_EVENT_QUEUE_SIZE;
```

Comment:
If `MAX_EVENT_QUEUE_SIZE` is a power of 2, we can apply optimization technique which converts to `event_queue.start &= MAX_EVENT_QUEUE_SIZE -1`, resulting in faster execution.

Reply:
See https://stackoverflow.com/questions/10527581/why-must-a-ring-buffer-size-be-a-power-of-2